### PR TITLE
For FNX-22339: UI components for recently added bookmarks

### PR DESCRIPTION
**Pull request by eliserichards**
For [FNX-22339](https://jira.mozilla.com/browse/FNX-22339)
Designs are [here](https://www.figma.com/file/am9EJGg6ow29PXWkc9jcOH/MR2.1?node-id=11%3A396)

UI components for recently added bookmarks on the homescreen. This is broken out of https://github.com/mozilla-mobile/fenix/pull/19835 so it's not so large. Diagram for how these are broken down: https://docs.google.com/drawings/d/1oiHqRl9f_LwsLbQ98zUBjChyDn5xm7us-iDF4rqhRak/edit?usp=sharing

- Header with "show all" button
- Bookmark item
- "Component" to house the items (contains `recent_bookmarks_list`)
- New styles & dimens to match the designs

### Screenshots

**header with show all button**
image

**button padding (accessibility)**
image

**Bookmark item with image, title, and subtitle**
image  image

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

File: app/src/main/res/layout/component_recent_bookmarks.xml
Comment (jonalmeida) [inline app/src/main/res/layout/component_recent_bookmarks.xml:13]: Shouldn't the contents of `recent_bookmarks_header` be here (above the `RecyclerView`)? Is there a situation where you can have a bookmarks header without the list?

You can use `<include layout="" />` if you want to keep some separation of concerns between the header and this instead of in-lining.

File: app/src/main/res/values/styles.xml
Comment (gabrielluong) [inline app/src/main/res/values/styles.xml:684]: You probably want to find out the dark theme color for this.

File: app/src/main/res/values/dimens.xml
Comment (gabrielluong) [inline app/src/main/res/values/dimens.xml:104]: I am kinda curious by seeing all these new `dimen` variables when we should be adding them in general. I have seen us add variables when it needs to be used programmatically through our kotlin files and when it's reusable across layouts and components. However, looking at how small this `dimens.xml` file is, I am wondering how we should plan to go about this here and in the future?

My immediate thought is that this might be useful if we were responsive and implementing for different device sizes, but historically, we have not done this.

File: app/src/main/res/values/dimens.xml
Comment (gabrielluong) [inline app/src/main/res/values/dimens.xml:115]: We can also create a new grouping `<!-- Recent bookmarks -->` to move towards a per feature component grouping.

**Review by gabrielluong — COMMENTED** [review]
I only looked at the header. You probably want another review for the bookmarks component.

File: app/src/main/res/layout/component_recent_bookmarks.xml
Comment (eliserichards) [reply app/src/main/res/layout/component_recent_bookmarks.xml:13]: Great point! I want to keep them separate for now in case we decide to do a placeholder component (like collections), so I think `<include layout..>` would be a good solution.

File: app/src/main/res/values/strings.xml
Comment (eliserichards) [reply app/src/main/res/values/strings.xml:56]: I wanted them near Home so there wasn't any confusion, but that totally makes sense 👍

File: app/src/main/res/values/styles.xml
Comment (eliserichards) [reply app/src/main/res/values/styles.xml:407]: It's not the `primaryText` color.. I'll have to see if one exists for `photonDarkGrey05` and if not I'll set one up

Comment (eliserichards) [reply app/src/main/res/values/styles.xml:407]: looks like "Jump back in" tabs is using a pre-defined color as well (`<color name="home_show_all_button_text">@color/photonDarkGrey05</color>`). I'm going to keep this as-is for now and fix them together in a separate PR when we get the Dark values

File: app/src/main/res/values/dimens.xml
Comment (eliserichards) [reply app/src/main/res/values/dimens.xml:104]: Not sure. I'll keep these here for now.

File: app/src/main/res/values/styles.xml
Comment (eliserichards) [reply app/src/main/res/values/styles.xml:413]: see below

File: app/src/main/res/layout/component_recent_bookmarks.xml
Comment (eliserichards) [reply app/src/main/res/layout/component_recent_bookmarks.xml:28]: Those aren't needed (at least by ktlint) in xml files, but I'll add it.

File: app/src/main/res/layout/recent_bookmarks_header.xml
Comment (eliserichards) [reply app/src/main/res/layout/recent_bookmarks_header.xml:1]: This can be in a different PR. There are a number of things that my components are going to add that are missing from first glance in the "Jump back in" pr and I'd rather get this landed first then deal with that.

Comment (gabrielluong) [reply app/src/main/res/layout/recent_bookmarks_header.xml:1]: In that case, I think you can duplicate https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/res/layout/recent_tabs_header.xml more or less here. It will help you cut back on the number of variables and really simplify your layout.

File: app/src/main/res/values/styles.xml
Comment (eliserichards) [reply app/src/main/res/values/styles.xml]: I got the dark theme colors from Nicole. Fixed it here and [filed a new issue](https://jira.mozilla.com/browse/FNX-22747) to add them to "Jump back in"

File: app/src/main/res/layout/recent_bookmarks_header.xml
Comment (eliserichards) [reply app/src/main/res/layout/recent_bookmarks_header.xml:1]: I'd prefer to keep it as-is

File: app/src/main/res/color/primary_state_button_text_color.xml
Comment (gabrielluong) [inline app/src/main/res/color/primary_state_button_text_color.xml:5]: I feel this change was in response to the dark theme feedback. I am not entirely sure since we can't really test its usage since it's not being used. However, if this is to adjust for dark theme, we actually do this quite differently without selectors and there's `nightly/colors.xml`

Comment (eliserichards) [reply app/src/main/res/color/primary_state_button_text_color.xml:5]: This is being used in another PR. I just want to get this landed.

Comment (eliserichards) [reply app/src/main/res/color/primary_state_button_text_color.xml:5]: To be clear, there is additional work needed to add dark themes for "Jump back in" as well. This doesn't have to happen in this PR.

Comment (gabrielluong) [reply app/src/main/res/color/primary_state_button_text_color.xml:5]: I guess this is what makes the PR and review a bit weird since it's just xml.

Comment (eliserichards) [reply app/src/main/res/color/primary_state_button_text_color.xml:5]: I believe I made it clear why this was broken out. You can see the link to the main PR in the description

Comment (gabrielluong) [reply app/src/main/res/color/primary_state_button_text_color.xml:5]: I am wondering what the additional work needed for the dark theme header since we already use `?primaryText` in `Header20TextStyle`. I think that's probably even more reasons we should just not have 2 headers and have to maintain both.

Comment (eliserichards) [reply app/src/main/res/color/primary_state_button_text_color.xml:5]: Yes, and this will happen in another PR. Not relevant for this.

File: app/src/main/res/color/recent_bookmark_item_text_color.xml
Comment (gabrielluong) [inline app/src/main/res/color/recent_bookmark_item_text_color.xml:5]: There's nothing wrong with using `<selector>` for the different text color state between light and dark theme. I personally would be interested in getting feedback from everyone on whether or not we should be moving towards doing it this way or just adding the color variables in `colors.xml` and `values-night/colors.xml` (https://searchfox.org/mozilla-mobile/search?q=home_show_all_button_text&path= for example).

File: app/src/main/res/values/styles.xml
Comment (gabrielluong) [reply app/src/main/res/values/styles.xml]: Ah yes, we just added `home_show_all_button_text`. It is using `photoLightGrey50` for the dark theme in the "Show all" button - see https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/res/values-night/colors.xml#99 (note the `values-nightly/colors.xml`). I checked with Nicole to about header colours since I didn't see this respond before. I still hold that we probably want to keep the `textColor` as `?primaryText` in these `BodyXXTextStyle` since it would be strange to have "12" be different from the rest and we can just add the `textColor` locally.

Comment (gabrielluong) [reply app/src/main/res/values/styles.xml]: The "Jump back in" header actually also uses the correct dark theme color with using `style="@style/Header20TextStyle"`. In our existing color variables and component system, `?primaryText` maps to https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/res/values-night/colors.xml#7 which maps to https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/res/values/colors.xml#80 which is `photonLightGrey05`. I will close the Jira issue since there's no change necessary.

You might be interested in reusing `home_show_all_button_text` since we anticipated the need for a generic "Show all" button text color.

File: app/src/main/res/values/styles.xml
Comment (jonalmeida) [inline app/src/main/res/values/styles.xml:666]: In general, we should avoid adding these kinds of styles in future work, it's a strange precedence that exists in the code base because, the way we're using it in Fenix, doesn't seem to add any value from just setting these one-time used styles directly in the layout files.

Nothing to change here for now, just food for thought.

**Review by jonalmeida — APPROVED** [review]

File: app/src/main/res/values/styles.xml
Comment (gabrielluong) [inline app/src/main/res/values/styles.xml:407]: Sorry, I am gonna blocked this on one change here, which might've been missed in the comments https://github.com/mozilla-mobile/fenix/pull/19953#discussion_r650204484.

```suggestion
        <item name="android:textColor">@color/primary_state_title_text_color</item>
```

**Review by gabrielluong — DISMISSED** [review]

File: app/src/main/res/values/styles.xml
Comment (eliserichards) [reply app/src/main/res/values/styles.xml:666]: That's good to know. We definitely don't want to make `styles` so huge that it becomes unusable

File: app/src/main/res/values/styles.xml
Comment (eliserichards) [reply app/src/main/res/values/styles.xml:407]: This is happening in the other PR, and I can't do that until this is merged.

Comment (eliserichards) [reply app/src/main/res/values/styles.xml:407]: Also, this is not supposed to be ?primaryText color, it's a different value

Comment (gabrielluong) [reply app/src/main/res/values/styles.xml:407]: I think there's some misunderstanding here. We want this to be `?primaryText` here since all the other BodyXXTextStyle uses `?primaryText`. 

Since this can `Body12TextStyle` is extended in `recentBookmarkItemSubTitleText`, we can just move  `<item name="android:textColor">@color/primary_state_title_text_color</item>` over there.

**Pull request merged**
